### ISSUE TITLE
JWT token generation is generating an invalid JWT issuer using {account} value

### ISSUE THREAD

**Issue by davlee1972** [snowflakedb/gosnowflake#1033]
The {account} value is being used to generate both the host uri and jwt issuer. It doesn't work if the account contains ".privatelink"

See discussion here including debug logging:
https://github.com/apache/arrow-adbc/issues/1422

The Snowflake PYTHON data connector does handle this properly and the same logic should exist in this GO driver..

** ".privatelink" is stripped from account when used to generate a JWT token issuer in PYTHON **
https://github.com/snowflakedb/snowflake-connector-python/blob/main/test/integ/test_key_pair_authentication.py
```
@pytest.mark.parametrize(
    "input_account,expected_account",
    [
        ("s3testaccount.global", "S3TESTACCOUNT.GLOBAL"),
        ("acct-with-dashes", "ACCT-WITH-DASHES"),
        ("testaccount.extra", "TESTACCOUNT"),
        ("testaccount-user.global", "TESTACCOUNT"),
        ("normalaccount", "NORMALACCOUNT"),
    ],
)
```
** Here's the python logic which strip ".privatelink" from the account for the JWT issuer, but preserves it for the HOST URI **
https://github.com/snowflakedb/snowflake-connector-python/blob/main/src/snowflake/connector/util_text.py

```
def construct_hostname(region: str | None, account: str) -> str:
    """Constructs hostname from region and account."""
    if region == "us-west-2":
        region = ""
    if region:
        if account.find(".") > 0:
            account = account[0 : account.find(".")]
        host = f"{account}.{region}.snowflakecomputing.com"
    else:
        host = f"{account}.snowflakecomputing.com"
    return host


def parse_account(account):
    url_parts = account.split(".")
    # if this condition is true, then we have some extra
    # stuff in the account field.
    if len(url_parts) > 1:
        if url_parts[1] == "global":
            # remove external ID from account
            parsed_account = url_parts[0][0 : url_parts[0].rfind("-")]
        else:
            # remove region subdomain
            parsed_account = url_parts[0]
    else:
        parsed_account = account

    return parsed_account

```
Please answer these questions before submitting your issue. 
In order to accurately debug the issue this information is required. Thanks!

1. What version of GO driver are you using?
v1.6.22
   
2. What operating system and processor architecture are you using?
Both Linux and Windows.. Intel Processors.
   
3. What version of GO are you using?
Unknown. Using a Python wrapper.

4.Server version:* E.g. 1.90.1
8.2

5. What did you do?

Running snowflake Go driver using python adbc driver.

6. What did you expect to see?

Successful connection

7. Can you set logging to DEBUG and collect the logs?

Yes
   
8. What is your Snowflake account identifier, if any? (Optional)

**Comment by sfc-gh-dszmolka** [comment]
hi and thank you for submitting this issue. So having only very few details about your situation (region, `gosnowflake` version used, etc and also no DEBUG available at this moment), tried to reproduce the issue using
1. deployment: AWS (EU Frankfurt)
2. driver version: `gosnowflake` v1.7.2

Configured privatelink, then used [keypair.go](https://github.com/snowflakedb/gosnowflake/blob/master/cmd/keypair/keypair.go) example script to attempt keypair authentication . 
```
$ env | grep SNOWFLAKE
SNOWFLAKE_TEST_USE=admin
SNOWFLAKE_TEST_ACCOUNT=MYACCOUNT.eu-central-1.privatelink <<< correct account format for privatelink connectivity
SNOWFLAKE_TEST_PRIVATE_KEY=/home/admin/go/rsa_key_snowflake.p8

$ go run cmd/keypair/keypair.go 
INFO[0000]log.go:216 gosnowflake.(*defaultLogger).Info OpenWithConfig                               
DEBU[0000]log.go:148 gosnowflake.(*defaultLogger).Debugf OCSP Cache Server for Privatelink: http://ocsp.MYACCOUNT.eu-central-1.privatelink.snowflakecomputing.com/ocsp_response_cache.json 
DEBU[0000]log.go:148 gosnowflake.(*defaultLogger).Debugf OCSP Retry URL for Privatelink: http://ocsp.MYACCOUNT.eu-central-1.privatelink.snowflakecomputing.com/retry/%v/%v 
INFO[0000]log.go:154 gosnowflake.(*defaultLogger).Infof Authenticating via SNOWFLAKE_JWT             
INFO[0000]auth.go:353 gosnowflake.authenticate PARAMS for Auth: &map[], https, MYACCOUNT.eu-central-1.privatelink.snowflakecomputing.com, 443, 5m0s, SNOWFLAKE_JWT 
INFO[0000]log.go:154 gosnowflake.(*defaultLogger).Infof full URL: https://MYACCOUNT.eu-central-1.privatelink.snowflakecomputing.com:443/session/v1/login-request?requestId=aa95fd11-[..]-a59c8b72c511&request_guid=66e7fc84-[..]-9fda274672ff 
..
DEBU[0000]rows.go:78 gosnowflake.(*snowflakeRows).Close Rows.Close                                    LOG_SESSION_ID=6415[..]314
Congrats! You have successfully run SELECT 1 with Snowflake DB!
```

looks to be working thus JWT is valid (otherwise could not authenticate)

Environment details could be important too (deployment, etc) and if there's a runnable reproduction Go program (using only `gosnowflake` without ADBC) which when run, leads to the issue, that would be great. Thank you in advance!

I see this originally started as a ADBC issue and it is entirely possible it lies within `gosnowflake` after all, but first I would like to be able to reproduce it without ADBC to get evidence of the issue lurking indeed in `gosnowflake` and not ABDC.

In the meantime I'm trying to verify the JWT generated from the Snowflake Go driver.

**Comment by davlee1972** [comment]
The go driver I'm using is 1.6.22

I can't log out the JWT iss but I'm pretty sure it contains .privatelink which is a problem.

If I use account without privatelink and also pass in the host uri with privatelink then the driver works.

394300 | JWT_TOKEN_INVALID_USER_IN_ISSUER | The user name specified in the issuer does not exist in the Snowflake account. For possible solutions, see Common Errors and Solutions.

**Comment by sfc-gh-dszmolka** [comment]
Is it possible for you to test with the latest gosnowflake and using something similar to `keypair.go` (in the example programs under `cmd/keypair` in this repo) . Just to be sure to use a very simple repro (which I used) and also to be sure you're not hitting a bug which might have been fixed since 1.6.22 (couple releases since then...)

to log the JWT generated I simply modified `auth.go`'s `prepareJWTToken` :
```go
tokenString, err := token.SignedString(config.PrivateKey)
logger.Debug(tokenString) // this line was added
```

I really feel this is not an issue in `gosnowflake`'s current versions but would be great to make sure.

**Comment by sfc-gh-dszmolka** [comment]
closing this issue as has been inactive for a while now, and it does not see coming from `gosnowflake` as in the reproduction we seem to correctly generate the JWT.

Please verify the JWT per the above comment and if you can reproduce the issue using the pure `gosnowflake` library without ADBC, it would be really helpful if you could please provide the reproduction steps/script as for now, it does not reproduce without ADBC.
